ICU-21734 Add missing overrides

stampho opened the pull request:
<!--
Thank you for your pull request!

Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license agreement (CLA) before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html
-->

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21734
- [x] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

jefgen commented on icu4c/source/common/unicode/brkiter.h:
```diff
@@ -146,7 +146,7 @@ class U_COMMON_API BreakIterator : public UObject {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
Why is this commented out? Should it be removed entirely perhaps?

stampho replied:
I've commented out because it is already pure virtual in the base class. I didn't know if I should remove this because the docs above. I saw getDynamicClassID commented out somewhere else in the code and thought this might be the proper way to keep the docs.
I will give a try with override = 0; thanks.
No problem, I will figure out how to handle the mistakenly landed suggestions :)

jefgen commented on icu4c/source/common/unicode/chariter.h:
```diff
@@ -142,7 +142,7 @@ class U_COMMON_API ForwardCharacterIterator : public UObject {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
Same question about this being commented out.

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const override = 0;

jefgen commented on icu4c/source/i18n/casetrn.h:
```diff
@@ -58,7 +58,7 @@ class CaseMapTransliterator : public Transliterator {
      * @return a copy of the object.
      */
-    virtual CaseMapTransliterator* clone() const = 0;
+    //virtual CaseMapTransliterator* clone() const = 0;
```
Same question about this being commented out.

FrankYFTang replied:
virtual CaseMapTransliterator* clone() const override = 0;

jefgen commented on icu4c/source/i18n/csrmbcs.h:
```diff
@@ -71,9 +71,9 @@ class CharsetRecog_mbcs : public CharsetRecognizer {
-    const char *getName() const = 0;
-    const char *getLanguage() const = 0;
-    UBool match(InputText* input, CharsetMatch *results) const = 0;
+    //const char *getName() const = 0;
```
Same question about these 3 lines being commented out.

FrankYFTang replied:
const char *getName() const override = 0;
    const char *getLanguage() const override = 0;
    UBool match(InputText* input, CharsetMatch *results) const override = 0;

jefgen commented on icu4c/source/i18n/unicode/coll.h:
```diff
@@ -864,7 +864,7 @@ class U_I18N_API Collator : public UObject {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
Same question about this being commented out.

FrankYFTang replied:
Notice with the line uncomment, the sublass of Collator are require to implment getDynamicClassID or it will get build/compile time error w/o but if we comment out this line, the superclass version of getDynamicClassID in UObject will kick in and the compiler won't caught such requirement during the compile/build time.

jefgen replied:
Ah, just to make sure I understand -- are you saying that by making `getDynamicClassID` a pure virtual function in the `Collator` class, this will force any subclasses of Collator to implement it? (And by commenting it out here, this now would permit fallback to the superclass `getDynamicClassID` implementation, which isn't what we'd want).

FrankYFTang replied:
sorry, I probably make a mistake while I wrote that comment. It looks like the = 0 line won't do what I though it will.

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const  override = 0;

jefgen commented on icu4c/source/i18n/unicode/datefmt.h:
```diff
@@ -229,13 +229,13 @@ class U_I18N_API DateFormat : public Format {
      * @stable ICU 2.0
      */
-    virtual DateFormat* clone() const = 0;
+    //virtual DateFormat* clone() const = 0;
```
Same question about this being commented out.

FrankYFTang replied:
virtual DateFormat* clone() const override = 0;

FrankYFTang commented on icu4c/source/common/unicode/unifilt.h:
```diff
@@ -77,7 +78,7 @@ class U_COMMON_API UnicodeFilter : public UnicodeFunctor, public UnicodeMatcher
      * @stable ICU 2.4
      */
-    virtual UnicodeFilter* clone() const = 0;
+    virtual UnicodeFilter* clone() const override { UPRV_UNREACHABLE; return nullptr; }
```
This change a compilation time assertion/error to a run time error and I am concern about this.

FrankYFTang replied:
virtual UnicodeFilter* clone() const override = 0;

stampho replied:
I did this because the base class is cast to UnicodeFilter* somewhere in the code. I will try the override = 0; I hope it will work.

FrankYFTang commented on icu4c/source/common/unicode/unifunct.h:
```diff
@@ -100,7 +100,7 @@ class U_COMMON_API UnicodeFunctor : public UObject {
      * @stable ICU 2.4
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
Why comment out this code?

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const override = 0;

FrankYFTang commented on icu4c/source/i18n/csrmbcs.h:
```diff
@@ -121,17 +121,13 @@ class CharsetRecog_euc : public CharsetRecog_mbcs
 public:
     virtual ~CharsetRecog_euc();
 
-    const char *getName() const = 0;
```
I have concern of these three delete

FrankYFTang replied:
const char *getName() const override = 0;

FrankYFTang commented on icu4c/source/i18n/csrsbcs.h:
```diff
@@ -74,64 +74,62 @@ class CharsetRecog_sbcs : public CharsetRecognizer
 public:
     CharsetRecog_sbcs();
     virtual ~CharsetRecog_sbcs();
-    virtual const char *getName() const = 0;
```
I have concern of theset two delete lines

FrankYFTang replied:
virtual const char *getName() const override = 0;

FrankYFTang commented on icu4c/source/i18n/csrucode.h:
```diff
@@ -33,12 +33,12 @@ class CharsetRecog_Unicode : public CharsetRecognizer
      * @see com.ibm.icu.text.CharsetRecognizer#getName()
      */
-    const char* getName() const = 0;
+    //const char* getName() const = 0;
```
why do you comment out this line?

FrankYFTang replied:
const char* getName() const override = 0;

FrankYFTang commented on icu4c/source/i18n/csrucode.h:
```diff
@@ -33,12 +33,12 @@ class CharsetRecog_Unicode : public CharsetRecognizer
      * @see com.ibm.icu.text.CharsetRecognizer#match(com.ibm.icu.text.CharsetDetector)
      */
-    UBool match(InputText* textIn, CharsetMatch *results) const = 0;
+    //UBool match(InputText* textIn, CharsetMatch *results) const = 0;
```
same issue of comment out line

FrankYFTang replied:
UBool match(InputText* textIn, CharsetMatch *results) const  override = 0;

FrankYFTang commented on icu4c/source/i18n/csrucode.h:
```diff
@@ -72,34 +72,32 @@ class CharsetRecog_UTF_32 : public CharsetRecog_Unicode
 
     virtual ~CharsetRecog_UTF_32();
 
-    const char* getName() const = 0;
```
why is this line got removed.

FrankYFTang replied:
const char* getName() const override = 0;

FrankYFTang commented on icu4c/source/i18n/unicode/basictz.h:
```diff
@@ -49,7 +50,7 @@ class U_I18N_API BasicTimeZone: public TimeZone {
      * @stable ICU 3.8
      */
-    virtual BasicTimeZone* clone() const = 0;
+    virtual BasicTimeZone* clone() const override { UPRV_UNREACHABLE; return nullptr; }
```
This change a build/compile time error/assertion to a run time assertion/error and I am concern about this.

FrankYFTang replied:
virtual BasicTimeZone* clone() const override = 0;

FrankYFTang commented on icu4c/source/i18n/unicode/calendar.h:
```diff
@@ -1341,7 +1341,7 @@ class U_I18N_API Calendar : public UObject {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
This change the requirement for the subclass to implement the getDynamicClassID() and remove the compile/build time error if the subclass does not do so but push to run time assertion/error and I am concern about it.

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const override = 0;

FrankYFTang commented on icu4c/source/i18n/unicode/numfmt.h:
```diff
@@ -266,15 +267,15 @@ class U_I18N_API NumberFormat : public Format {
      * @stable ICU 2.0
      */
-    virtual NumberFormat* clone() const = 0;
+    virtual NumberFormat* clone() const override { UPRV_UNREACHABLE; return nullptr; }
```
This push the compile/build time error checking of the subclass of NumberFormat not implementing a clone method to a runtime assertion / error check and I am concern about this.

FrankYFTang replied:
virtual NumberFormat* clone() const override = 0;

FrankYFTang commented on icu4c/source/i18n/unicode/numfmt.h:
```diff
@@ -1077,7 +1078,7 @@ class U_I18N_API NumberFormat : public Format {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
This should not be removed. Once removed, the build/compile time error reporting of the subclass of NumberFormat w/o implement getDynamicClassID would be removed and may have a subclass w/o implement such.

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const override = 0;

FrankYFTang commented on icu4c/source/i18n/unicode/timezone.h:
```diff
@@ -790,7 +790,7 @@ class U_I18N_API TimeZone : public UObject {
      * @stable ICU 2.0
      */
-    virtual UClassID getDynamicClassID(void) const = 0;
+    //virtual UClassID getDynamicClassID(void) const = 0;
```
This removal will remove the build/compile time checking of subclass implement getDynamicClassID.

FrankYFTang replied:
virtual UClassID getDynamicClassID(void) const override = 0;

FrankYFTang reviewed (changes requested):
Thanks for working on this. We really appreciate it. In genearl it looks good except some lines you commented out or adding run time check to avoid build time failure. 
We would also like to add build bot in github action to build with the right warning check once / with this PR so in the future we will not regress on it. Since this PR is very big. I would suggest we do the following:
1. Remove the parts we have issues w/ from this PR so we can approve the non-contraversial changes in this PR and land it soon.
2. Create a second PR to have those contraversial changes Jeff and I commented on and let's work on that smaller PR to address the open issue together. In this way, when you have some changes on those part jeff and I  do not need to through 175 files to look double check to validate everytime you make a push. 
3. I will try to craft a PR for the build bot change to valid currently it break the build w/o you chage in this PR and once we validate that on github we can merge that part into your second PR 
4.

FrankYFTang commented:
Sorry, my ealier comments had some misunderstanding about the  getDynamicClassID()

FrankYFTang reviewed (changes requested):
also you need to change the code inside test/* and make sure the "make tests" will compile all the code.

stampho commented:
hi @FrankYFTang and @jefgen , 

1. I have updated this PR and left out the questionable parts
2. The questionable parts are in https://github.com/unicode-org/icu/pull/1846
3. I also added overrides in the tests: https://github.com/unicode-org/icu/pull/1847

I don't know how to add you to the pull-requests as a reviewer, probably I don't have rights for it.

jefgen commented:
Thanks for updating this PR, and for making the other 2 PRs! :)

It looks like the "WarningsAsError" build bots are complaining about the two header files `unifilt.h` and `unifunct.h`:

```
./unicode/unifilt.h:80:28: error: 'clone' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
    virtual UnicodeFilter* clone() const = 0;
                           ^
./unicode/unifunct.h:54:29: note: overridden virtual function is here
    virtual UnicodeFunctor* clone() const = 0;
                            ^
```

But the change to add `override` to these two `clone()` methods is now in the other PR here: [ICU-21734 Add missing overrides to pure virtual methods #1846](https://github.com/unicode-org/icu/pull/1846)

However, I'm thinking that in order to keep the build bots happy (and to avoid checking in changes that make them fail), perhaps we could omit the changes to the headers `uniflt.h` and `unifunct.h` in this PR and do them as part of the other PR (#1846) after this is merged?

stampho commented:
> However, I'm thinking that in order to keep the build bots happy (and to avoid checking in changes that make them fail), perhaps we could omit the changes to the headers `uniflt.h` and `unifunct.h` in this PR and do them as part of the other PR (#1846) after this is merged?

I don't really understand how this will fix the buildbot since both headers are included in some other files and those will still cause warnings. I doesn't even change unifunct.h in this PR.

Anyway, I did what you suggested and moved unifilt.h changes from here to #1846 . I hope it helps.

stampho commented:
Hi @jefgen , I've tried to move all the common source files to #1846. Now, its commit message is a bit misleading but seems to be working.

jefgen commented:
> Hi @jefgen , I've tried to move all the common source files to #1846. Now, its commit message is a bit misleading but seems to be working.

Thank you!
Thanks for moving the changes around. It looks like the "WarningsAsErrors" build bot is okay with the changes now: https://github.com/unicode-org/icu/pull/1825/checks?check_run_id=3525667678

jefgen reviewed (approved):
Thank you for all of these changes. 👍

FrankYFTang reviewed (approved):
LGTM. Thanks
